modules/weko-workflow/weko_workflow/utils.py

```diff
@@ -1469,3 +1471,22 @@ def create_files_url(root_url, record_id, filename):
         root_url,
         record_id,
         filename)
+
+
+def save_indexs_publish_doi(item_id):
+    """Save indexs publish doi.
+
+    :param item_id: Item id.
+
+    :return:
+    """
+    rm = RecordMetadata.query.filter_by(id=item_id).first()
+    indexs = rm.json['path']
+    list_indexs = []
+    for idx in indexs:
+        for index in idx.split('/'):
+            if index not in list_indexs:
+                list_indexs.append(index)
+                index_database = Index.query.filter_by(id=index).first()
+                index_database.public_state = True
+                db.session.commit()
```

Comment:
To avoid access to DB many times, just add a flag to check whether need to update or not, if so then do update.
Sample:
```
    rm = RecordMetadata.query.filter_by(id=item_id).first()
    index_id_list = rm.json['path']
    updated_index_ids = []
    update_db = False
    for index_ids in index_id_list:
        for index_id in index_ids.split('/'):
            if index_id not in updated_index_ids:
                updated_index_ids.append(index_id)
                update_db = True
                index = Index.query.filter_by(id=index_id).first()
                index.public_state = True
    if update_db:
        db.session.commit()
```